single_kernel_mongo/utils/mongo_connection.py

```diff
@@ -333,3 +357,261 @@ def is_any_sync(rs_status: dict[str, Any]) -> bool:
+        """Returns a list of databases using the given shard as a primary shard.
+
+        In Sharded MongoDB clusters, mongos selects the primary shard when creating a new database
+        by picking the shard in the cluster that has the least amount of data. This means that:
+        1. There can be multiple primary shards in a cluster.
+        2. Until there is data written to the cluster there is effectively no primary shard.
+        """
+        config_db = self.client["config"]
+        if "databases" not in config_db.list_collection_names():
+            logger.info("No data written to sharded cluster yet.")
+            return None
+
+        databases_collection = config_db["databases"]
+        if databases_collection is None:
+            return None
+
+        return databases_collection.distinct("_id", {"primary": primary_shard})
+
+    def _move_primary(self, databases_to_move: list[str], old_primary: str) -> None:
+        """Moves all the provided databases to a new primary.
+
+        Raises:
+            NotEnoughSpaceError, ConfigurationError, OperationFailure
+        """
+        for database_name in databases_to_move:
+            db_size = self.get_db_size(database_name, old_primary)
+            new_shard, avail_space = self.get_shard_with_most_available_space(
+                shard_to_ignore=old_primary
+            )
+            if db_size > avail_space:
+                no_space_on_new_primary = (
+                    f"Cannot move primary for database: {database_name}, new shard: {new_shard}",
+                    f"does not have enough space. {db_size} > {avail_space}",
+                )
+                logger.error(no_space_on_new_primary)
+                raise NotEnoughSpaceError(no_space_on_new_primary)
+
+            # From MongoDB Docs: After starting movePrimary, do not perform any read or write
+            # operations against any unsharded collection in that database until the command
+            # completes.
+            logger.info(
+                "Moving primary on %s database to new primary: %s. Do NOT write to %s database.",
+                database_name,
+                new_shard,
+                database_name,
+            )
+            # This command does not return until MongoDB completes moving all data. This can take
+            # a long time.
+            self.client.admin.command("movePrimary", database_name, to=new_shard)
+            logger.info(
+                "Successfully moved primary on %s database to new primary: %s",
+                database_name,
+                new_shard,
+            )
+
+    def get_db_size(self, database_name, primary_shard) -> int:
+        """Returns the size of a DB on a given shard in bytes."""
+        database = self.client[database_name]
+        db_stats = database.command("dbStats")
+
+        # sharded databases are spread across multiple shards, find the amount of storage used on
+        # the primary shard
```

Comment:
I assume this function was copied from the original?

I am forgetful of the logic of why we only want to find the amount of storage used on the primary shard.

Regardless maybe the function name should be changed to get_db_size_on_primary_shard

Reply:
Yes it was copied.
We need to find the amount of storage used because we cannot move primary to a shard that doesn't have enough available space so we try to find the one which has the most available space and if it's still to small, we raise an error so the user can act accordingly.

I can rename it yes

Reply:
Can you please add the context of what you just described in the docstring of this function